gp-templates/helper-functions.php

```diff
@@ -69,123 +69,150 @@ function gp_prepare_translation_textarea( $text ) {
+			$term = substr( $term, 0, -2 );
+			$suffixes[] = 'fe';
+			$suffixes[] = 'ves';
 		} elseif ( 'an' === substr( $term, -2 ) ) {
-			$terms[] = substr( $term, 0, -2 ) . 'en';
+			$term = substr( $term, 0, -2 );
+			$suffixes[] = 'an';
+			$suffixes[] = 'en';
 		} else {
-			$terms[] = $term . 'es';
-			$terms[] = $term . 'ed';
-			$terms[] = $term . 'ing';
+			$suffixes[] = 's';
+			$suffixes[] = 'es';
+			$suffixes[] = 'ed';
+			$suffixes[] = 'ing';
 		}
 
-		$glossary_entries_terms[ $key ] = implode( '|', $terms );
+		$glossary_entries_suffixes[ $term ] = $suffixes;
 	}
 
 	// Sort by length in descending order.
-	uasort(
-		$glossary_entries_terms,
+	uksort(
+		$glossary_entries_suffixes,
 		function( $a, $b ) {
 			return mb_strlen( $b ) <=> mb_strlen( $a );
 		}
 	);
 
-	return $glossary_entries_terms;
+	return $glossary_entries_suffixes;
 }
 
 /**
  * Add markup to a translation original to identify the glossary terms.
  *
  * @param GP_Translation $translation            A GP Translation object.
  * @param GP_Glossary    $glossary               A GP Glossary object.
- * @param array          $glossary_entries_terms A list of terms to highligh.
  *
  * @return obj The marked up translation entry.
  */
-function map_glossary_entries_to_translation_originals( $translation, $glossary, $glossary_entries_terms = null ) {
-	$glossary_entries = $glossary->get_entries();
-	if ( empty( $glossary_entries ) ) {
-		return $translation;
-	}
-
-	if ( null === $glossary_entries_terms || ! is_array( $glossary_entries_terms ) ) {
-		$glossary_entries_terms = gp_sort_glossary_entries_terms( $glossary_entries );
-	}
+function map_glossary_entries_to_translation_originals( $translation, $glossary ) {
+	static $terms_search, $glossary_entries_reference, $glossary_entries, $cached_glossary;
+	if ( isset( $terms_search ) && isset( $cached_glossary ) && $cached_glossary === $glossary->id ) {
+		if ( ! $terms_search ) {
+			return $translation;
+		}
+	} else {
+		// Build our glossary search.
+		$glossary_entries = $glossary->get_entries();
+		$cached_glossary = $glossary->id;
+		if ( empty( $glossary_entries ) ) {
+			$terms_search = false;
+			return $translation;
+		}
 
-	$glossary_entries_terms_array = array();
-	$glossary_term_reference      = array();
+		$glossary_entries_suffixes = gp_glossary_add_suffixes( $glossary_entries );
 
-	// Since glossary entries terms have been returned as an array of strings, we need to break it up and create back references.
-	foreach ( $glossary_entries_terms as $i => $terms ) {
-		// Split the terms in to an array and make sure they are all lower case for later use.
-		$term_list = explode( '|', $terms );
-		$term_list = array_map( 'strtolower', $term_list );
+		$glossary_entries_reference = array();
+		foreach ( $glossary_entries as $id => $value ) {
+			$term = strtolower( $value->term );
+			if ( ! isset( $glossary_entries_reference[ $term ] ) ) {
+				$glossary_entries_reference[ $term ] = $id;
```

Comment:
```suggestion
				$glossary_entries_reference[ $term ] = array( $id );
```

Was there a reason why you didn't use an array from the start here? If it's an array we could remove all the `(array)` type casting, which would make the code more readable IMO.

Reply:
It was an attempt to save memory but I tried it and it didn't make a difference, so I'm changing these to arrays, thanks!